src/lightning_app/components/auto_scaler.py

```diff
@@ -0,0 +1,575 @@
+                pending_requests_per_running_or_pending_work = metrics["pending_requests"] / (
+                    replicas + metrics["pending_works"]
+                )
+
+                # upscale
+                if pending_requests_per_running_or_pending_work >= max_requests_per_work:
+                    return replicas + 1
+
+                # downscale
+                if pending_requests_per_running_or_pending_work < min_requests_per_work:
+                    return replicas - 1
+
+                return replicas
+
+
+        app = L.LightningApp(
+            MyAutoScaler(
+                MyPythonServer,
+                min_replicas=1,
+                max_replicas=8,
+                autoscale_interval=10,
+                max_batch_size=8,  # for auto batching
+                timeout_batching=2,  # for auto batching
+            )
+        )
+    """
+
+    def __init__(
+        self,
+        work_cls: Type[LightningWork],
+        min_replicas: int = 1,
+        max_replicas: int = 4,
+        autoscale_interval: int = 1 * 10,
+        max_batch_size: int = 8,
+        timeout_batching: float = 2,
+        endpoint: str = None,
+        input_type: type = Dict,
+        output_type: type = Dict,
+        *work_args: Any,
+        **work_kwargs: Any,
+    ) -> None:
+        super().__init__()
+        self.num_replicas = 0
+        self._work_registry = {}
+
+        self._work_cls = work_cls
+        self._work_args = work_args
+        self._work_kwargs = work_kwargs
+
+        self._input_type = input_type
+        self._output_type = output_type
+        self.autoscale_interval = autoscale_interval
+        self.max_batch_size = max_batch_size
+
+        if max_replicas < min_replicas:
+            raise ValueError(
+                f"`max_replicas={max_replicas}` must be less than or equal to `min_replicas={min_replicas}`."
+            )
+        self.max_replicas = max_replicas
+        self.min_replicas = min_replicas
+        self._last_autoscale = time.time()
+        self.fake_trigger = 0
+
+        endpoint = endpoint or "api/predict"
+        self.load_balancer = _LoadBalancer(
+            input_type=self._input_type,
+            output_type=self._output_type,
+            endpoint=endpoint,
+            max_batch_size=max_batch_size,
+            timeout_batching=timeout_batching,
+            cache_calls=True,
+            parallel=True,
+        )
+        for _ in range(min_replicas):
+            work = self.create_work()
+            self.add_work(work)
+
+    @property
+    def workers(self) -> List[LightningWork]:
+        return [self.get_work(i) for i in range(self.num_replicas)]
+
+    def create_work(self) -> LightningWork:
+        """Replicates a LightningWork instance with args and kwargs provided via ``__init__``."""
+        # TODO: Remove `start_with_flow=False` for faster initialization on the cloud
+        return self._work_cls(*self._work_args, **self._work_kwargs, start_with_flow=False)
```

Comment:
**For follow-up.** The limitation here is that we would need to not have random work names, which should be ok if just the initial works have non-random names. Any that be deleted and re-created would need random names I think - from what I remember you can't remove and re-create a work of the same name during the apps life.